### ISSUE TITLE
amp-list cause line-break that cannot be reversed with CSS

### ISSUE THREAD

**Issue by jaygray0919** [ampproject/amphtml#23675]
Our goal is to implement a tool-tip similar to this Washington Post example:
  https://www.washingtonpost.com/sports/2019/03/19/blake-bortles-signs-with-rams-one-year-deal-back-up-jared-goff/?utm_term=.6657442a70fc

Here are three scenarios to illustrate our issue:
  https://dailyfood.io/amp/tooltip/question.html
A quick explanation of the 3 scenarios:
1  inline but *no* `amp-list` (to prove that `inline` keeps multiple `<div>`s together)
2  *one* `amp-list` (hierarchical); `amp-lightbox` for *EACH* `item`
3  multiple `amp-list` (_one_ _list_ _per_ `item`; multiple items); *one* `amp-lightbox` template

AMP page rendering adds CSS which, somehow-somewhere, forces a line break in scenario-3. Candidates:
```
i-amphtml-element
i-amphtml-layout
i-amphtml-fill-content
i-amphtml-replaced-content
```
Another possibility, from AMP documentation:
```
amp-list div[role="list"] {
    display: grid;
    grid-gap: 0.5em;
  }
```

For us, this is a `severity-2` issue because we have a work-around (scenario-2). But scenario-2 dictates significant page-bloat due to "one `amp-lightbox` per item", and the `amp-mustache` item processing (#open /close) adds complexity and bloat.

Bottom line: we cannot use multiple `amp-list`s in a single paragraph, as is effectively achieved in the WaPo example:
  https://www.washingtonpost.com/sports/2019/03/19/blake-bortles-signs-with-rams-one-year-deal-back-up-jared-goff/?utm_term=.6657442a70fc

Cathy Zhu @cathyxz looked at this issue and suggested we copy @CrystalOnScript Crystal Lambert on this thread.
Cathy thought Crystal might have a template/example that uses `amp-bind`, where we could bind an `amp-list` _item_ to a `<span>` element (an inline element that would allow us to use _items_ from multiple `amp-list`s in a single `<p>`). We then would link the `amp-list` _item_ to an `amp-lightbox` template.

However, an **ideal** **solution** would fix scenario-3, without requiring the overhead of an `amp-bind` step. An `amp-bind` step makes the structure less portable/reusable among different documents.

**Comment by jaygray0919** [comment]
@cathyxz anything we can do to help here Cathy? This is important (but not critical; not a sev-1) for our first release.

**Comment by cathyxz** [comment]
The soonest time I can take an in-depth look at this will probably be our next fixit week. @sebastianbenz do you by chance have any ideas on existing solutions for implementing tooltips? If I recall correctly, some portions of @jaygray0919 's feature requests seemed similar to the seatmap example that @kul3r4 built based on amp-list and amp-pan-zoom a while ago, but I can't find a live example of that using amp-list anymore.

**Comment by jaygray0919** [comment]
we started the tooltip journey with a @sebastianbenz draft (an inline selector).
the ONLY problem is the inability (using CSS) to override a line-break generated by `amp-runtime` when using `amp-list`. (we've actually embedded the `amp-runtime-CSS` on a page [like `amp.dev`] and tried to over-ride using a 'keep-together' CSS statement - it fails).
The only time you see the issue is when 2+ `amp-list`s are used for term-substitution in a paragraph.
in contrast, if we use only one (hierarchical array) `amp-list`, the problem does not occur.
but we have libraries of single-item `amp-list`s (e.g. values from Wikipedia) that are used dynamically.
so, if the WaPo player football stats (used as a target example in a previous explanation) were in player-specific `amp-list`s, their paragraphs also would line-break on each player-tooltip.
this is so frustrating we considered using a sidebar as a tooltip. we have a prototype, but it's too much work to customize the amount of space for tooltips with irregular size.
separately, @sebastianbenz did help us with integration of `amp-list` and an SVG, but there is nothing there that we could adapt to the discussion above.
the bottom line (IOHO): eliminate the line-break generated by `amp-runtime` for an `amp-list` so that it can be used with an an HTML element like `<span>`

**Comment by cathyxz** [comment]
Okay, rereading this, I think your issue boils down to needing an amp-list inside an amp-lightbox to bind to different sources of data. 

You can bind the `src` attribute of `<amp-list>` to an `<amp-state>`, whose datasource can be either inlined or generated dynamically via the `src` attribute. The syntax would look something like this: 

```
<amp-list
      width="396"
      height="800"
      layout="fixed"
      items="."
      [src]="categories[selected]">
      <ul>
      <template type="amp-mustache">
          <li>{{title}}</li>
      </template>
    </ul>
    </amp-list>
```

The `[src]` attribute here can reference a subsection of the `<amp-state>` which has a JSON structure. A small example of binding amp-list src to amp-state can be found here: https://codepen.io/cathyxz/pen/jONaKJY

/cc @nainar in case this actually boils down to needing a tooltip FR.

**Comment by nainar** [comment]
Hi @jaygray0919 sorry for potentially explaining yourself again. 

Taking a look at what the WaPo article you linked to, it's not a tooltip in the sense of what I think a tooltip is

What I think a tooltip is: https://getbootstrap.com/docs/4.3/components/tooltips/

What you want is a tappable context to launch a lightbox modal right?

**Comment by jaygray0919** [comment]
@cathyxz Am away until next week and will work your proposal then. @nainar Tooltips historically have used title="" and mouseover. Extensions build on that original idea. However, mouseover works poorly in mobile. @sebastianbenz recognized this and designed an amp-tooltip that was a selector. We added the amp-lightbox feature to add information. WaPo tooltip also is based on a selector and lightbox and works well on mobile. Conventional bootstrap++ tooltips do NOT work well on mobile (I leave experiments to you on this point). IOHO the WaPo design is the best design for mobile tooltip and is our target. The issue we have is a conflict with amp-list. Hopefully @cathyxz solution will work. Will report back end of next week.

**Comment by jaygray0919** [comment]
@cathyxz Cathy we've studied your suggestion.
IOHO it is equivalent, effectively, to using `amp-mustache` selections (i.e. # and /) to harvest a value from an `amp-list`.

Here is our configuration:
We have a REST server with multiple items.
Our REST server is modeled on PubChem (https://pubchemdocs.ncbi.nlm.nih.gov/pug-rest).
Here is an AMP query that harvests PubChem data (https://afdsi.org/test/amp_list_tree_pubchem/).

Here is our use-case:
Using an inline element (like `<span>`) harvest a single item from our REST server.
The required structure is identical to the WaPo tooltip example:
 

- a single item (a football player) and properties for that item that are harvested by `amp-mustache`.

The major point: multiple tooltips are in a single HTML block element (e.g. `<p>`).

The problem:
`amp-list`, or `amp-list` + `amp-lightbox` injects a line break.
Therefore we cannot use `<span>` as a container for a single `<amp-list>`.
If WaPo used AMP for its Sports Page tooltip, they would have the same problem.
Reason: data for each tooltip is in a separate file, and multiple tooltips are part of a single block structure `<p>`.

Our work-around:
Use one `<amp-list>` with an hierarchical structure (very different from the single item format for our REST endpoint!) and `amp-mustache` selections to harvest the target value.
This design is not impeded by the line-break generated by `amp-list`.
But, it corrupts our REST endpoint.

Our proposal:
Replace the AMP runtime function that injects the line-break with a non-breaking element.
We have tried multiple CSS methods to over-ride the breaking element - all fail.

Bottom line: eliminate AMP runtime elements that force a line-break.

cc @nainar

**Comment by jaygray0919** [comment]
Any chance of getting this issue fixed in the upcoming #wg-ui-and-a11y sprint?

**Comment by wassgha** [comment]
Hey @caroqliu do you mind looking into this please? @jaygray0919 I'm not sure I see a tooltip on the page you linked to, could you please post a screenshot of what you're trying to achieve?

**Comment by jaygray0919** [comment]
Dang; the WaPo has deprecated their tooltip with player statistics. Previously the player's name was in a highlighted box that, when clicked, revealed their statistics (i.e. the name 'Blake Bortles' was in a box that opened in a lightbox). In any event, the inability to override amp-runtime-injected-css is causing the line break. We are asking for the ability to insure that the injected controls do not cause a line break. See our example for illustrations.

**Comment by nainar** [comment]
@jaygray0919 I am afraid the current proposal you are suggesting will cause breaking changes in the AMP codebase. This is something we can do with a new version of `<amp-list>` but not with the current one. 

I think the wider tooltip FR is a good one to have. It's something that is hard to do with AMP currently and we should create one for AMP. However, given the presence of a workaround and the fact that a large part of the team will be working on Bento in 2020, I don't reasonably think we will get to this feature soon. 

Having said that contributions are welcome and if you are willing to contribute someone on the WG will be happy to work with you here.

**Comment by jaygray0919** [comment]
thanks for the response @nainar . Our current thinking is: generate the tooltip inline. The single-structure `amp-list` + `amp-mustache` parsing is difficult to maintain - and really does not offer an advantage v. conventional server-side, inline HTML. Please evaluate our request as you consider/develop `amp-list-0.2` - we'll revisit our tooltip structure if supported in that release.

**Comment by jaygray0919** [comment]
@nainar I see that amp-list v2 is under development: https://github.com/ampproject/amphtml/issues/16507 and https://amp.dev/community/roadmap/. Would it be appropriate to solve our problem as part of that project?

**Comment by jaygray0919** [comment]
@caroqliu Caroline - anything we can do here to help solve this problem? Hoping that, at a minimum, it's on the amp-list v2 development plan.

**Comment by samouri** [comment]
Is this the same issue as https://github.com/ampproject/amphtml/issues/31532?
If not, what would be most helpful for us to solve it is if you created the smallest html file you can that demonstrates the issue.

**Comment by jaygray0919** [comment]
No, it's not the same as #31532 But I will do as requested. Previously we constructed 5 use-cases for @choumx to illustrate the issue. The use-cases ranged from simple to complex. Will rehost those examples over the weekend, recheck them, and then republish. Will notify you Monday/Tuesday when they are ready for review. Thank you for looking into this.

**Comment by jaygray0919** [comment]
FYI here is one dialog with Will: https://github.com/ampproject/amphtml/issues/17509#issuecomment-413701362. Here is another: https://github.com/ampproject/amphtml/issues/23672. Those are the use-cases that we'll restore, recheck and republish.

**Comment by jaygray0919** [comment]
@samouri Thanks for helping us here.

Below are two examples that support the presentation:

  1 [example-4-40](https://afdsi.com/samouri-amp/v4-40)

  * Summary: 
    * hierarchical JSON
    * CSS "keep-together" **works**


  2 [example-6-40](https://afdsi.com/samouri-amp/v6-40)

  * Summary
    * one amp-state-json per item
    * 2+ items in a paragraph
    * CSS "keep-together" **fails**
    * amp-template creates line breaks that break up a paragraph


Our goal: support this logical structure
```
<p>
<span>{{term-from-amp-list-23}}</span> author-text <span>{{term-from-amp-list-14}}</span> author-text <span>{{term-from-amp-list-6}} author-text.</span>
</p>
```
  * where term-from-amp-list-<em>NN</em> can be reused on many amp-pages
  * where all amp-pages that use term-from-amp-list-<em>NN</em> are updated at the same time (when server-side data is refreshed based on the last sporting event)

Imagine this use-case:
  * A database of sport-players and their statistics.
  * Players are referenced in past and current discussions.
  * Discussions occur on different amp-pages of an amp-site.
  * Discussion text about a player on a date is immutable (it is what an author wrote at that time).
  * Player statistics are current as-of the last event, not as-of the date of the discussion.
  * A player's statistics are available anywhere (on the amp-site) where the player is referenced.
  * Authored-discussions about a player reference several players (and their statistics) in the same paragraph.

`amp-list/state/mustache` does not support this use-case.

Instead, we must use a page-specific `amp-list/state/mustache` structure that includes the data about all players discussed on the amp-page (an hierachical JSON structure that is parsed by player identity).

Then we have to `mustache/open/close` the items used on the amp-page.

In contrast, this structure
```
<template id=23><span>{{term-from-amp-list-23}}</template> author-text <template id=14><span>{{term-from-amp-list-24}}</template> ...
```
would be more flexible (easier to build and reuse) ...

where `amp-list/state` is at the bottom of the page (we inject list/state between <\/body> and <\/html> so we always "know where it is") Sebastian Benz taught us to do this.

Also, injecting list/state at the bottom of the page allows us to generalize `height=` (vs. having to be more specific/exact if list/state is adjacent to \<template>).

---

If you can make the change to enable amp/state/list/template `in-line`, we plan to use it **as soon** as it is available.

At this time, our work-around is the hierarchical JSON approach as illustrated in `example v4-40`

Ask me questions and I'll get right back to you.

/jay

**Comment by jaygray0919** [comment]
@samouri anything i can do to help here? any unansered questions?

**Comment by jaygray0919** [comment]
@samouri Also discussed the issue here: https://github.com/ampproject/amphtml/issues/10230

**Comment by jaygray0919** [comment]
@samouri anything i can do to help fix this issue?

**Comment by jaygray0919** [comment]
@samouri Here is an example of an inline tooltip - the type we want to implement using `amp-list`

https://www.singlelunch.com/2020/12/28/why-im-lukewarm-on-graph-neural-networks/

See the second sentence:

        A regular graph <tooltip> is just another way to think about a particular flavor

The inability to use `amp-list` as part of an inline element (e.g. `<span>`) makes this impossible.

IOHO, based on looking at code using the `amp-optimizer`, you need to switch from `block element` to `inline element`

Can this be implemented soon?
